docs: added info about the pipeline + links to other steps repos (#72)
alercebroker/kubernetes_pipeline: PIPELINE.md

# An in-depth guide to the ALeRCE pipeline

Every step works as a "event driven microservice", which listens to a topic and must receive one or more messages in order to perform their operations. The pipeline steps are developed under the [APF Framework](https://github.com/alercebroker/APF). This framework gives some strong guidelines on the structure of the step, in addition of providing some hooks which facilitates the parsing of the information received by the step and the further deployment as well.  

## Pipeline Diagram (detailed)

*insert image here*

## Steps specification

### Sorting Hat

> The Sorting Hat is a gateway step which assign an ALeRCE ID (aid) to every incoming alert. This `aid` can be newly generated or an already existing one, depending on the result of the crossmatch performed on the object database. 

Reviewer: ```suggestion
The Sorting Hat is a gateway step which assigns an ALeRCE ID (aid) to every incoming alert. This `aid` can be newly generated or an already existing one, depending on the result of the crossmatch performed on the object database. 
```

> At the present time, this step must be deployed with different with different setting to work with the different surveys availible. So you might find two different instances of sorting hat deployed.

Reviewer: ```suggestion
At the present time, this step must be deployed once per survey, with different settings to work with the given survey. So you might find more than one instance of sorting hat deployed.
```

#### Input Schema

```python
{
    "type": "record",
    "doc": "Multi stream alert of any telescope/survey",
    "name": "alerce.alert",
    "fields": [
        {"name": "oid", "type": "string"},
        {"name": "tid", "type": "string"},
        {"name": "pid", "type": "long"},
        {"name": "candid", "type": ["long", "string"]},
        {"name": "mjd", "type": "double"},
        {"name": "fid", "type": "int"},
        {"name": "ra", "type": "double"},
        {"name": "dec", "type": "double"},
        {"name": "rb", "type": ["null", "float"]},

>         {"name": "rbversion", "type": ["null", "string"]},

Reviewer: Eliminados en la nueva versión

        {"name": "mag", "type": "float"},
        {"name": "e_mag", "type": "float"},

>         {"name": "rfid", "type": ["null", "int"]},

Reviewer: Eliminado en la nueva versión

Reviewer: eso mismo

        {"name": "pid", "type": "long"},
        {"name": "candid", "type": ["long", "string"]},
        {"name": "mjd", "type": "double"},
        {"name": "fid", "type": "int"},
        {"name": "ra", "type": "double"},
        {"name": "dec", "type": "double"},
        {"name": "mag", "type": "float"},
        {"name": "e_mag", "type": "float"},
        {"name": "isdiffpos", "type": "int"},
        {"name": "e_ra", "type": "float"},
        {"name": "e_dec", "type": "float"},
        {
            "name": "extra_fields",
            "type": EXTRA_FIELDS,
        },
        {"name": "aid", "type": "string"},
        {"name": "stamps", "type": STAMPS},
    ],
}
```

Given the **STAMPS** type:

```python
{
    "type": "record",
    "name": "stamps",
    "fields": [
        {"name": "science", "type": ["null", "bytes"], "default": None},
        {"name": "template", "type": ["null", "bytes"], "default": None},
        {"name": "difference", "type": ["null", "bytes"], "default": None},
    ],
}
```

and the **EXTRA_FIELDS** type:

```python
{
    "type": "map",
    "values": ["null", "int", "float", "string", "bytes"],
    "default": {},
}
```

### Stamp Classifier Step

> Applies a machine learning model to determine the classification of an object based on a image of the former, called stamp.

Reviewer: ```suggestion
Applies a machine learning model to determine the classification of an alert based on its stamps. This should have an instance per survey, since the ML models are survey dependent.
```

#### Output Information

This step doesn't publish to any Kafka topic related to the pipeline.

### Previous Candidates Step

> Parses the `extra_fields` of the detections to retrieve the binary field `prv_candidates` if the alert comes from ZTF. Otherwise, it does nothing. Then, this step returns a list with the processed objects, including their `aid`, previous candidates, detections without the `extra_fields` field and their non_detections.

Reviewer: ```suggestion
Parses the `extra_fields` of the detections to retrieve the binary field `prv_candidates` if the alert comes from ZTF. This will add to the main alert the previous detections and (possibly) non-detections. For other surveys, only the main alert is passed among the detections and the non-detections are empty. 

This step returns a list with the processed objects, including their `aid`, detections (main alert and previous detections), and non-detections.
```

            {
              "name": "extra_fields",
              "type": {
                "default": {},
                "type": "map",
                "values": ["null", "int", "float", "string", "bytes"]
              }
            },
            {
              "name": "aid",
              "type": "string"
            }
          ]
        },
        "default": []
      }
    },
    {
      "name": "non_detections",
      "type": {
        "type": "array",
        "items": {
          "type": "record",
          "name": "non_detection",
          "fields": [
            {
              "name": "aid",
              "type": "string"
            },
            {
              "name": "tid",
              "type": "string"
            },
            {
              "name": "oid",
              "type": "string"
            },
            {
              "name": "mjd",
              "type": "double"
            },
            {
              "name": "fid",
              "type": "int"
            },
            {
              "name": "diffmaglim",
              "type": "int"
            }
          ]
        }
      },
      "default": []
    }
  ]
}
```

### Lightcurve Step 

> This step retrieves from the database and then returns the entire lightcurve of an object.
> It merges the repeated objects within the same batch of messages.

Reviewer: ```suggestion
This step retrieves from the database and then returns all stored detections and non-detections of an object, together with the new detections/non-detections coming from the previous step.

Repeated objects within the same batch of messages are merged.
```

> #### Output Schema

Reviewer: Acá me falta el `has_stamp`. Quizás sería mejor decir que este tiene exactamente el mismo esquema que el paso anterior

Author: Sugerencia tomada

                  "type": "float"
              },
              {
                  "name": "extra_fields",
                  "type": {
                  "default": {},
                  "type": "map",
                  "values": ["null", "int", "float", "string", "bytes"]
                  }
              }
            ]
        },
        "default": []
      }
    },
    {
      "name": "non_detections",
      "type": {
        "type": "array",
        "items": {
          "type": "record",
          "name": "non_detection",
          "fields": [
            {
              "name": "aid",
              "type": "string"
            },
            {
              "name": "tid",
              "type": "string"
            },
            {
              "name": "oid",
              "type": "string"
            },
            {
              "name": "mjd",
              "type": "double"
            },
            {
              "name": "fid",
              "type": "int"
            },
            {
              "name": "diffmaglim",
              "type": "int"
            }
          ]
        }
      },
      "default": []
    }
  ]
}
```

### Correction Step

> Process alerts and applies correction to pass from difference magnitude to apparent magnitude. This includes the previous detections coming from the ZTF alerts.

Reviewer: ```suggestion
Process detections and applies correction to pass from difference magnitude to apparent magnitude.

It also calculates the object mean RA and declination based on the position of its detections.
```

#### Output Schema

```python
{
  "type": "record",
  "doc": "Previous candidates schema with new alert and previous detections and non detections",
  "name": "prv_candidates",
  "fields": [

>     {
>       "name": "aid",
>       "type": "string"
>     },

Reviewer: Esto es nuevo, pero habría que añadirlo

```suggestion
    {
      "name": "aid",
      "type": "string"
    },
    {
      "name": "meanra",
      "type": "float"
    },
    {
      "name": "meandec",
      "type": "float"
    },
```

            {"name": "dec", "type": "double"},
            {"name": "e_dec", "type": "double"},
            {"name": "mag", "type": "float"},
            {"name": "e_mag", "type": "float"},
            {"name": "isdiffpos", "type": "int"},
            {"name": "rb", "type": ["float", "null"]},
            {"name": "rbversion", "type": ["string", "null"]},
            {
                "name": "extra_fields",
                "type": {
                    "type": "map",
                    "values": ["string", "int", "null", "float", "boolean", "double"],
                },
            },
        ],
    },
}
```

the `NON_DETECTIONS` type:

```python
{
    "type": "array",
    "items": {
        "name": "non_detections_record",
        "type": "record",
        "fields": [
            {"name": "tid", "type": "string"},
            {"name": "oid", "type": "string"},
            {"name": "mjd", "type": "double"},
            {"name": "diffmaglim", "type": "float"},
            {"name": "fid", "type": "int"},
        ],
    },
}
```

and the `XMATCH` type:

```python
{
    "type": "map",
    "values": {"type": "map", "values": ["string", "float", "null", "int"]},

> }

Reviewer 2: Me hice la misma pregunta cuando revise el esquema, vasicamente dice que los xmatch son diccionarios con valores de cualquier tipo. No?

> }

Reviewer: Exacto, dice que puede ser cualquier cosa, pero me tinca que no es tan así

> }

Reviewer 2: Hasta ahora no sabemos como tiene que ser... Podriamos revisar en la db si hay un patron?

> }

Reviewer: Mi duda es si el README del step está al día, porque si es más de un catálogo creo que perdimos, pero si es sólo allwise podemos definir esto mejor

> }

Reviewer 2: Sorry. Es el esquema que esta en el step de xmatch. Esta al dia hasta donde sabemos.
